refactor: fix warnings from clang-tidy-20 and bitcoin-tidy

ryanofsky opened the pull request:
Warnings are described in commit messages and were reported by Sjors in https://github.com/bitcoin/bitcoin/pull/31802#issuecomment-2858241771

ryanofsky committed:
clang-tidy: Fix bugprone-move-forwarding-reference error

/ci_container_base/include/mp/type-interface.h:47:75: error: forwarding reference passed to std::move(), which may unexpectedly cause lvalues to be moved; use std::forward() instead [bugprone-move-forwarding-reference,-warnings-as-errors]
https://cirrus-ci.com/task/6187773452877824?logs=ci#L4712

Error was caused by mp/type-interface.h CustomBuildField field using std::move
instead of std::forward.

This commit:

- Adds std::forward and std::move several places to preserve lvalue/rvalue
  status.

- Defines a FunctionTraits::Fwd<...>(...) helper which does same thing as
  std::forward except it takes parameter number instead of a parameter type so
  generated code doesn't need as verbose as std::forward calls would be.

- Changes C++ code generator to pass arguments as `M0::Fwd<1>(arg1)` instead of
  `arg1` in generated code. This change can be verified by diffing a generated
  file like build/src/ipc/capnp/mining.capnp.proxy-client.c++ in the bitcoin
  build before and after this change.

Unlike the previous commit which resolved bugprone-move-forwarding-reference
errors by just switching std::move to std::forward, this commit required more
changes because just switching from std::move to std::forward in the
type-interface.h CustomBuildField function would lead to another error in the
CustomMakeProxyServer function there which is expecting an rvalue, not an
lvalue.

That error could have alternately been fixed by changing CustomMakeProxyServer
to accept lvalues and copy the shared_ptr. But this would be slightly less
efficient and it is better to resolve the problem at the root and just use
perfect forwarding everywhere, all the way up the call stack. This required the
changes to the code generator and clientInvoke described above.

ryanofsky committed:
clang-tidy: Suppress bitcoin-nontrivial-threadlocal error

/ci_container_base/include/mp/proxy-io.h:637:1: error: Variable with non-trivial destructor cannot be thread_local. [bitcoin-nontrivial-threadlocal,-warnings-as-errors]
https://cirrus-ci.com/task/6187773452877824?logs=ci#L4720

ryanofsky committed:
clang-tidy: Fix bugprone-move-forwarding-reference errors

Reported by Sjors Provoost <sjors@sprovoost.nl>
https://github.com/bitcoin/bitcoin/pull/31802#issuecomment-2858241771

/ci_container_base/include/mp/proxy-io.h:252:16: error: forwarding reference passed to std::move(), which may unexpectedly cause lvalues to be moved; use std::forward() instead [bugprone-move-forwarding-reference,-warnings-as-errors]
/ci_container_base/include/mp/proxy-io.h:336:18: error: forwarding reference passed to std::move(), which may unexpectedly cause lvalues to be moved; use std::forward() instead [bugprone-move-forwarding-reference,-warnings-as-errors]
/ci_container_base/include/mp/util.h:186:12: error: forwarding reference passed to std::move(), which may unexpectedly cause lvalues to be moved; use std::forward() instead [bugprone-move-forwarding-reference,-warnings-as-errors]
https://cirrus-ci.com/task/6187773452877824?logs=ci#L4712

hebasto commented:
Concept ACK.

ryanofsky commented:
Rebased 8d8e9f65f1d9faaafe91ca90910c1f96da9d094d -> e6a71ae1e2ca2186fd464ddee19eaefac6fd567c ([`pr/tidy.1`](https://github.com/ryanofsky/libmultiprocess/commits/pr/tidy.1) -> [`pr/tidy.2`](https://github.com/ryanofsky/libmultiprocess/commits/pr/tidy.2), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/tidy.1-rebase..pr/tidy.2)) fixing typos and conflict with #165

Sjors commented:
I added the latest version of this to https://github.com/bitcoin/bitcoin/pull/31802 to see if CI is still happy.

ryanofsky commented:
Thanks @Sjors. Now that this git repo is a subtree and a bitcoin-core project, I would like to (1) stop my previous practice of merging PRs like this that do not have any ACKs and (2) try to encourage more review to happen earlier in the upstream PR's in this repository rather than later in the downstream PRs in the bitcoin repository.

So would you feel comfortable reviewing more and maybe giving an ACK along the lines of...

- Testing this PR and confirming it fixes the warnings
- Looking at the code and confirming it only disables one `bitcoin-tidy` warning and resolves several `bugprone-move-forwarding-reference` warnings by changing `std::move` to `std::forward` in some places and changing lvalue references to rvalue references other places

... or however you might actually review it? I am happy to answer any questions that could help clarify things. The second commit should be the only complicated one here, and I can break it down. It is:

- Adding `std::forward` and `std::move` several places to preserve lvalue/rvalue status.
- Defining a `FunctionTraits::Fwd<...>(...)` helper which does same thing as `std::forward` except it takes parameter number instead of a parameter type so generated code doesn't need as verbose as `std::forward` calls would be.
- Changing C++ code generator to pass arguments as `M0::Fwd<1>(arg1)` instead of `arg1` in generated code. You could verify this by diffing a generated file like `build/src/ipc/capnp/mining.capnp.proxy-client.c++` in the bitcoin build before and after this change.

Maybe I should add some of these explanations in code comments or commit messages too. Would welcome any feedback.

Sjors commented:
> Maybe I should add some of these explanations in code comments

This might help more people understand the code.

ryanofsky commented:
re: https://github.com/bitcoin-core/libmultiprocess/pull/172#issuecomment-2887085223

> This might help more people understand the code.

Thanks, added new information above to code comments and commit messages.

Updated a5bff37ad8f5630a738eb1b0bfa60cf01bedbcdf -> 57a65b854664c13c9801788eeaf71bdb3e597c81 ([`pr/tidy.3`](https://github.com/ryanofsky/libmultiprocess/commits/pr/tidy.3) -> [`pr/tidy.4`](https://github.com/ryanofsky/libmultiprocess/commits/pr/tidy.4), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/tidy.3..pr/tidy.4)) updating these comments

Sjors commented:
ACK 34c9cee380e7276cf3f85f2ce56a293e57afd676

I studied the second commit more, along with some of `mp/gen.cpp`, thanks for the extra comments.

I wrote:

> I don't have an informed opinion on the last commit https://github.com/bitcoin-core/libmultiprocess/commit/a5bff37ad8f5630a738eb1b0bfa60cf01bedbcdf that suppresses bitcoin-nontrivial-threadlocal, but it has a comment with rationale.

The original `nontrivial-threadlocal` check was added in https://github.com/bitcoin/bitcoin/pull/30146 and reviewed by @maflcko and @hebasto, both already on this thread.

TheCharlatan commented on include/mp/proxy-io.h:
```diff
@@ -634,7 +634,10 @@ void ListenConnections(EventLoop& loop, int fd, InitImpl& init)
 
-extern thread_local ThreadContext g_thread_context;
+extern thread_local ThreadContext g_thread_context; // NOLINT(bitcoin-nontrivial-threadlocal)
+// Silence nonstandard bitcoin tidy error "Variable with non-trivial destructor
+// cannot be thread_local" which should not be a problem on modern platforms, and
+// could lead to a small memory leak at worst on older ones.
```
I've read through the history again of it again and came to the same conclusion.

ryanofsky replied:
re: https://github.com/bitcoin-core/libmultiprocess/pull/172#discussion_r2112039864

> My understanding is that we try to avoid it purely out of style/preference: [bitcoin/bitcoin#30095 (comment)](https://github.com/bitcoin/bitcoin/pull/30095#discussion_r1608289449). But no strong opinion.

Agree in general it's good to avoid `thread_local` destructors. Code will be more clear and explicit if it joins threads and deletes any state associated with them after they are joined, instead adding isolated `thread_local` variables that get initialized and cleaned up less deterministically.

But in this case, the libmultiprocess library needs to track state associated with threads that it doesn't have any control of.

In the bitcoin codebase, node, wallet, gui, and RPC threads are all able to make IPC calls, and the IPC implementation needs to track information about threads that make IPC calls so it can guarantee calls from the same local thread always run on the same remote thread, and guarantee that if the remote thread makes a callback it will run on the same local thread (otherwise things like recursive locks will not work). The IPC implementation also should be able to free state associated with threads when they are destroyed, so it uses a `thread_local` struct to do all these things.

I think keeping the bitcoin-tidy check but allowing use of a `thread_local` destructor here to detect when unmanaged external threads are destroyed is a reasonable fix. Other possible fixes not relying on `thread_local` might be possible too, but would involve other tradeoffs: they might rely on OS specific calls instead, or change the memory model to delete state when connections are closed not when threads are destroyed, or change the threading model to not always run calls from same IPC client thread on same IPC server thread. These fixes would add complexity other places, so I don't think would be preferable.

Sjors replied:
>  guarantee calls from the same local thread always run on the same remote thread, and guarantee that if the remote thread makes a callback it will run on the same local thread

I assume you mean that local has threads correspond one to one with remote threads, perhaps even sharing the same name, but they're not actually the same thread given that they're in a different process?

Sjors replied:
I'm curious to learn how non-multiprocess clients like #174 experience this one-to-one mapping.

ryanofsky replied:
re: https://github.com/bitcoin-core/libmultiprocess/pull/172#discussion_r2115146811

> I'm curious to learn how non-multiprocess clients like #174 experience this one-to-one mapping.

When you are using the libmultiprocess library as the IPC client, it forces you into the 1:1 threading model. The first time you make an IPC call from a local thread, the library will [automatically](https://github.com/bitcoin-core/libmultiprocess/blob/27c7e8e5a581b3c41330e758951251ef11807b11/include/mp/type-context.h#L31-L50) make a remote [ThreadMap.makeThread](https://github.com/bitcoin-core/libmultiprocess/blob/27c7e8e5a581b3c41330e758951251ef11807b11/include/mp/proxy.capnp#L47) call to execute the current IPC request and any future requests on.

But if you are using the Cap'n Proto interface directly from rust or c++ without libmultiprocess, you don't need to have a 1:1 threading model. You can create remote threads freely and make any IPC method call which takes a [`Context`](https://github.com/bitcoin-core/libmultiprocess/blob/27c7e8e5a581b3c41330e758951251ef11807b11/include/mp/proxy.capnp#L56-L65) argument run on whichever thread you specify in the [`Context.thread`](https://github.com/bitcoin-core/libmultiprocess/blob/27c7e8e5a581b3c41330e758951251ef11807b11/include/mp/proxy.capnp#L59-L60) field.

re: https://github.com/bitcoin-core/libmultiprocess/pull/172#discussion_r2115144124

> I assume you mean that local threads correspond one to one with remote threads, perhaps even sharing the same name, but they're not actually the same thread given that they're in a different process?

Yes that's right. The sentence is also trying to say if a local thread makes an IPC call, and during that IPC call the remote thread needs to call back into the local process, the callback will be executed on *literally* the same thread that made the first IPC call and is currently waiting for the response.

It's important for this to happen, for example, if the thread making the first IPC call locks a recursive mutex before the call. A callback from the server to the client will need to run on literally the same client thread to avoid deadlocking if it locks the mutex recursively.

TheCharlatan commented:
ACK 57a65b854664c13c9801788eeaf71bdb3e597c81

ryanofsky reviewed (commented):
Thanks for the reviews! Will try to merge this and update the subtree
